### PR TITLE
✨ SubnetSchema support

### LINKED ISSUE #4784: Subnet division enhacement
/kind feature

**Describe the solution you'd like**
By default, Cluster API AWS organizes subnets so that most of the VPC address space goes to private subnets. While this is true for most classic Kubernetes deployments, that may not work in some specific cases. 

We in Wargaming use Kuberentes to schedule arenas with gameservers using google agones. These arenas requires to be run on a node with a public IP address using hostPort. As pods with arenas are the main cluster workload, we need to dedicate most of VPC address space to the public subnets, instead of private.

In order to support this requirement, I would like to propose an additional configuration parameter in `AWSCluster.spec.network.vpc` - `SubnetSchema` with the following enum values:
- `PreferPrivate` - will be the default value and will implement the current logic
- `PreferPublic` - will have the reverse logic, where most of the VPC address space will be dedicated to public subnets
- `Equal` - will divide VPC address space equally between public and private subnets

### PR DESCRIPTION
**Pull request by krasoffski**
**What type of PR is this?**
This PR adds functionality related to Subnet schema type, more details can be found in original issue #4784.

Add one of the following kinds:
/kind feature

**What this PR does / why we need it**:
As I mentioned above, this PR adds ability to configure  preferred way for subnets division with priority to private as original behavior by default or public via `PreferPublic`.
For existing users nothing is changed because original behavior is not changed and set as default. But new parameter `PreferPublic`  allows to use provider in more specific cases like with Google Agones where subnet division to with public subnets priority fits better. 

**Which issue(s) this PR fixes**
Fixes #4784

**Special notes for your reviewer**:
My first PR to AWS provider =) 

**Checklist**:

- [x] squashed commits
- [x] includes documentation
- [x] includes [emojis](https://github.com/kubernetes-sigs/kubebuilder-release-tools?tab=readme-ov-file#kubebuilder-project-versioning)
- [x] adds unit tests
- [x] updates e2e tests

**Release note**:

```release-note
New  `AWSCluster.spec.network.vpc.SubnetSchema` parameter  which allows to change preferred way of subnet division to public and private. 
```

### CODE REVIEW

**Comment by krasoffski** [comment]
Despite the fact, PR marked as `size/XL` the code change is small and simple.  Rest of 90% of change is adding new unit tests.

**Comment by krasoffski** [comment]
Hi, @dlipovetsky and @richardcase
Please approve pipeline for this PR.

**Comment by krasoffski** [comment]
Hi, @dlipovetsky and @richardcase
Guys, please take a look.

**Comment by krasoffski** [comment]
Hi, @dlipovetsky and @richardcase
Could you please review this PR?
It is about 2 months since I created PR.

**Comment by richardcase** [comment]
Thank you for your patience @krasoffski. It's taken be a while to get back into things after a 6 week break away. 

/ok-to-test

File: api/v1beta1/types.go [hunk 1]
```diff
@@ -291,3 +291,14 @@ const (
 	// AmazonLinuxGPU is the AmazonLinux GPU AMI type.
 	AmazonLinuxGPU EKSAMILookupType = "AmazonLinuxGPU"
 )
+
+// SubnetSchemaType specifies how given network should be divided on subnets
+// in the VPC depending on the number of AZs.
+type SubnetSchemaType string
```
Comment (richardcase) [inline api/v1beta1/types.go:297]: API changes only need to go into the newest API version (i.e. v1beta2).  So these change here can be removed.

Comment (richardcase) [reply api/v1beta1/types.go:297]: There will be some changes required as a result with the "round trip fuzzy" tests. I can help with these if you run into issues.

File: api/v1beta2/network_types.go [hunk 2]
```diff
@@ -461,6 +461,15 @@ type VPCSpec struct {
 	// the API Server.
 	// +optional
 	ElasticIPPool *ElasticIPPool `json:"elasticIpPool,omitempty"`
+
+	// SubnetSchema specifies how CidrBlock should be divided on subnets in the VPC depending on the number of AZs.
+	// PreferPrivate - one private subnet for each AZ plus one other subnet that will be further sub-divided for the public subnets.
+	// PreferPublic - have the reverse logic of PreferPrivate, one public subnet for each AZ plus one other subnet
+	// that will be further sub-divided for the private subnets.
+	// Defaults to PreferPrivate
```
Comment (richardcase) [inline api/v1beta2/network_types.go:469]: It would be worth adding `// +optional` as well.

**Comment by krasoffski** [comment]
Thank you, @richardcase 

I will take a look for test problems and come back.

File: api/v1beta2/network_types.go (hunk 2 above)
Comment (krasoffski) [reply api/v1beta2/network_types.go:469]: Done

**Comment by krasoffski** [comment]
Hi, @richardcase 
I fixed comments and adjusted tests to the latest main.
Need confirmation from your side.

**Comment by krasoffski** [comment]
@richardcase 
May I ask a question.

Do these jobs have timeouts or how long it usually takes to finish?

I am asking because I see that one of them running more than an hour.
image

**Comment by richardcase** [comment]
Timeout is 4 hours: https://github.com/kubernetes-sigs/cluster-api-provider-aws/blob/main/Makefile#L163

**Comment by richardcase** [comment]
And the prow job has a 5 hour timeout: https://github.com/kubernetes/test-infra/blob/master/config/jobs/kubernetes-sigs/cluster-api-provider-aws/cluster-api-provider-aws-presubmits.yaml#L335

**Comment by krasoffski** [comment]
Thank you, @richardcase

**Review by alexander-demicev — APPROVED** [review]
thanks, the change looks good

/lgtm

**Comment by richardcase** [comment]
The e2e tests have passed. 

This change looks good to me, thanks for this @krasoffski .

/approve

**Comment by krasoffski** [comment]
Thank you, Guys.
I am very appreciate this.

**Pull request merged**